File: apple/TrainTimeWatch/Utilities/GeoUtils.swift
import Foundation
import CoreLocation

enum GeoUtils {
    /// Approximate distance in meters using flat-earth approximation (matches Garmin implementation)
    static func haversineDistance(from: CLLocationCoordinate2D, to: CLLocationCoordinate2D) -> Double {
        let dLat = (to.latitude - from.latitude) * 111000.0
        let dLon = (to.longitude - from.longitude) * 75700.0
        return (dLat * dLat + dLon * dLon).squareRoot()
    }

    /// Bearing in radians from one coordinate to another (great-circle, matches Garmin's calculateBearing)
    static func bearing(from: CLLocationCoordinate2D, to: CLLocationCoordinate2D) -> Double {
        let dLon = (to.longitude - from.longitude) * .pi / 180.0
        let lat1R = from.latitude * .pi / 180.0
        let lat2R = to.latitude * .pi / 180.0
        let y = sin(dLon) * cos(lat2R)
        let x = cos(lat1R) * sin(lat2R) - sin(lat1R) * cos(lat2R) * cos(dLon)
        return atan2(y, x)
    }

    /// Walk time in minutes at 83 m/min (matches Garmin's walkSpeed)
    static func walkMinutes(distanceMeters: Double) -> Double {
        distanceMeters / 83.0
    }

    /// Format walk info string matching Garmin's display
    /// When walkTimeSeconds is provided (from MKDirections), uses that instead of the 83 m/min estimate
    static func formatWalkInfo(distanceMeters: Double, walkTimeSeconds: Double? = nil) -> String {
        let walkMin: Int
        if let walkTime = walkTimeSeconds {
            walkMin = Int(walkTime / 60.0)
        } else {
            walkMin = Int(walkMinutes(distanceMeters: distanceMeters))
        }
        let timeStr = walkMin < 1 ? "<1 min" : "\(walkMin) min"
        return "\(timeStr) walk - \(Int(distanceMeters))m"
    }
}

File: android/core/src/main/kotlin/com/evanjt/traintime/domain/GeoUtils.kt
package com.evanjt.traintime.domain

import com.evanjt.traintime.Thresholds
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

object GeoUtils {
    // Flat-earth approximation, matches the Garmin and iOS implementations.
    fun haversineDistance(fromLat: Double, fromLon: Double, toLat: Double, toLon: Double): Double {
        val dLat = (toLat - fromLat) * 111000.0
        val dLon = (toLon - fromLon) * 75700.0
        return sqrt(dLat * dLat + dLon * dLon)
    }

    // Great-circle bearing in radians.
    fun bearing(fromLat: Double, fromLon: Double, toLat: Double, toLon: Double): Double {
        val dLon = Math.toRadians(toLon - fromLon)
        val lat1R = Math.toRadians(fromLat)
        val lat2R = Math.toRadians(toLat)
        val y = sin(dLon) * cos(lat2R)
        val x = cos(lat1R) * sin(lat2R) - sin(lat1R) * cos(lat2R) * cos(dLon)
        return atan2(y, x)
    }

    fun walkMinutes(distanceMeters: Double): Double = distanceMeters / Thresholds.WALK_SPEED

    fun formatWalkInfo(distanceMeters: Double, walkTimeSeconds: Double? = null): String {
        val walkMin = if (walkTimeSeconds != null) {
            (walkTimeSeconds / 60.0).toInt()
        } else {
            walkMinutes(distanceMeters).toInt()
        }
        val timeStr = if (walkMin < 1) "<1 min" else "$walkMin min"
        return "$timeStr walk - ${distanceMeters.toInt()}m"
    }
}
